treewide: Fix links in the docs and cleanup many boards (#22257)
RIOT-OS/RIOT: doc/guides/misc/openocd.md

---
title: OpenOCD
description: This page describes the process of using OpenOCD for flashing.
---

## Description

[OpenOCD](http://openocd.sourceforge.net/) (the open on-chip debugger) is an
open source tool for debugging and flashing microcontrollers. In RIOT we try
to use this tool for as many platforms as possible to reduce the overhead of
having to keep track of many different (and sometimes proprietary) tools.

## How to Build/Install

> ### Linux Ubuntu

Reviewer: ```suggestion
### Linux (Ubuntu)
```

---
title: OpenOCD
description: This page describes the process of using OpenOCD for flashing.
---

## Description

> [OpenOCD](http://openocd.sourceforge.net/) (the open on-chip debugger) is an open source tool for debugging and flashing microcontrollers. In RIOT we try to use this tool for as many platforms as possible to reduce the overhead of having to keep track of many different (and sometimes proprietary) tools.

Reviewer: ```suggestion
[OpenOCD](http://openocd.sourceforge.net/) (the open on-chip debugger) is an
open source tool for debugging and flashing microcontrollers. In RIOT we try
to use this tool for as many platforms as possible to reduce the overhead of
having to keep track of many different (and sometimes proprietary) tools.
```

## How to Build/Install

### Linux Ubuntu

> The following steps are for building and installing OpenOCD under Ubuntu. For other Linux distributions the steps have to slightly altered.

Reviewer: ```suggestion
The following steps are for building and installing OpenOCD under Ubuntu.
For other Linux distributions the steps have to slightly altered.
```

```shell
sudo apt install openocd
```

> 

Reviewer: ```suggestion

:::note
Ubuntu usually ships very old versions of OpenOCD, which lack support for
modern microcontrollers. In this case, you will have to build and install
OpenOCD yourself. Consider using the `latest` branch instead of a Release,
since the release cycle of OpenOCD is quite slow as well.
:::

```

### MacOS

1. Using [homebrew](http://brew.sh/) one can install the most recent version of OpenOCD by:

```shell
brew install [--HEAD] open-ocd
```

1. Test if everything worked out properly:

```shell
openocd --version
```

This should give you the version you have just build, probably something like `0.9.0-dev-xxxx`. Voila, that's it!

> 

Reviewer: ```suggestion

### Compiling OpenOCD

Please refer to the
[OpenOCD documentation](https://github.com/openocd-org/openocd#compiling-openocd)
for details about how to compile OpenOCD for your system.

```

#### Problems & Fixes

When using OpenOCD on Mac OS and you encounter an error similar to

```text
Error: libusb_claim_interface() failed with LIBUSB_ERROR_ACCESS
Error: unable to open ftdi device with vid 0403, pid 6010, description '*' and serial '*'
```

> make sure you are running ```make flash``` with ```sudo``` and that the system `ftdi` driver is unloaded:

Reviewer: ```suggestion
make sure you are running `make flash` with `sudo` and that the system `ftdi` driver is unloaded:
```

From [OpenOCD documentation](https://github.com/openocd-org/openocd/blob/master/README.macOS.md):

If you're using a USB adapter and have a driver kext matched to it, you will need to unload it prior to running OpenOCD. E.g. with Apple driver (OS X 10.9 or later) for FTDI run:

```shell
sudo kextunload -b com.apple.driver.AppleUSBFTDI
```

for FTDI vendor driver use:

```shell
sudo kextunload FTDIUSBSerialDriver.kext
```

### OpenOCD Always Return Error Even if Flashing Succeeded

The introduction of [#2897](https://github.com/RIOT-OS/RIOT/pull/2897)
always return a failed flashing even if it actually succeeds.

We recommend to install an OpenOCD version higher than `Open On-Chip
Debugger 0.9.0-dev-gbdd8312 (2016-01-12-18:06)`, if possible version 0.10
or higher. A reference to this issue can be found at [#6043](https://github.com/RIOT-OS/RIOT/issues/6043).

> 

Reviewer: ```suggestion
```

Not the case anymore.

### Flashing with ST-Link

If you run the last released version of OpenOCD (0.10) instead of building yourself, you will have to specify the version of your ST-Link.
The default is 2.1, if you are using an older adapter try to set

```bash
STLINK_VERSION=2
```

if you have the error `Error: open failed`.

The Makefile tells OpenOCD to do a reset on the board. If you have a simple ST-LINK adapter or the reset is not connected you can get the error: `Error: jtag status contains invalid mode value - communication failure` You can disable the reset via OpenOCD with the variable:

```bash
RSRT=none
```

This might require multiple retries or pushing the reset manually on the board just after starting the download.

> 

Reviewer: ```suggestion
```

Not the case anymore.

## Transports and Board Support

> To figure out if a board is supported by OpenOCD it's best to a look on the scripts that come with OpenOCD. On Ubuntu these are put into `/usr/local/share/openocd/scripts/`. RIOT uses these OpenOCD scripts via its own Makefiles. You can find them in the `./RIOT/makefiles/tools/openocd-adapters/` directory.

Reviewer: ```suggestion
To figure out if a board is supported by OpenOCD it's best to a look on the
scripts that come with OpenOCD. On Ubuntu these are put into
`/usr/local/share/openocd/scripts/`. RIOT uses these OpenOCD scripts via its
own Makefiles. You can find them in the
`RIOT/makefiles/tools/openocd-adapters/` directory.
```

> ## udev Rules

Reviewer: ```suggestion
## udev Rules and Unsupported Boards
```

Reviewer: This section is now somewhat redundant with the STM32 flashing guide, but that's okay I guess.

Reviewer: If you want to, you can remove this section:

```
If you have a recent OpenOCD version, you can check if your user is in the dialout and/or plugdev group and therefore has the appropriate access rights.

In either of the first two cases you have to compile the latest OpenOCD sources for your system and manually update the /etc/udev/rules.d/60-openocd.rules file with the file provided in the contrib subfolder of the OpenOCD repository. Please refer to the OpenOCD documentation for further information.
```

from the https://doc.riot-os.org/board_specific/stm32/ guide and refer to this subsection of the OpenOCD guide instead.

> When using OpenOCD on Linux and you encounter something similar to

Reviewer: ```suggestion
When using OpenOCD on Linux, your user might not have the appropriate access
rights to access the USB port directly. Furthermore, old versions of OpenOCD might
not have the support for newer programmers such as the ST-Link V3.
You might encounter an error similar to this:
```

```shell
Error: libusb_open() failed with LIBUSB_ERROR_ACCESS
Error: no device found
Error: unable to open ftdi device with vid 0403, pid 6010, description 'FITECO M3' and serial '*'
in procedure 'init'
```

chances are you have to configure udev rules to work with the device.
When built from source, OpenOCD provides examples in `contrib/99-openocd.rules`. Copy this
file to your udev configuration directory make sure your `vid` and `pid` are covered and

> adjust the `GROUP` to some group you are part of.

Reviewer: ```suggestion
If you have a recent OpenOCD version, you can check if your user is in the `dialout` and/or
`plugdev` group and therefore has the appropriate access rights.

If you have an old OpenOCD version, you have to compile the latest OpenOCD
sources for your system and manually update the
`/etc/udev/rules.d/60-openocd.rules` file with the file provided in the
`contrib` subfolder of the OpenOCD repository. Please refer to the OpenOCD
documentation for further information.
```

To reflect the new rules in the system

```shell
udevadm control --reload

> ```

Reviewer: ```suggestion
```

The udev rules stuff changes in OpenOCD, so I'd keep it at referring to the documentation.
